File: src/data/courses/dsa/stacks/stack-core-operations/implement-stack-operations/code/java/ImplementStackOperations.java
// ============================================================
// Implement Stack Operations
// ============================================================

import java.util.*;

// ============================================================
// Solution: Array-Based Stack
// All operations O(1) amortized
// ============================================================
class MyStack {
    private List<Integer> data = new ArrayList<>();

    public void push(int x) {
        data.add(x); // O(1) amortized
    }

    public int pop() {
        return data.remove(data.size() - 1); // O(1)
    }

    public int top() {
        return data.get(data.size() - 1); // O(1)
    }

    public boolean isEmpty() {
        return data.isEmpty(); // O(1)
    }

    public int size() {
        return data.size(); // O(1)
    }
}

File: src/data/courses/dsa/stacks/stack-core-operations/implement-stack-operations/code/python/implement_stack_operations.py
# ============================================================
# Implement Stack Operations
# ============================================================

# ============================================================
# Solution: Array-Based Stack
# All operations O(1) amortized
# ============================================================
class MyStack:
    def __init__(self):
        self.data = []

    def push(self, x: int) -> None:
        self.data.append(x)  # O(1) amortized

    def pop(self) -> int:
        return self.data.pop()  # O(1)

    def top(self) -> int:
        return self.data[-1]  # O(1)

    def is_empty(self) -> bool:
        return len(self.data) == 0  # O(1)

    def size(self) -> int:
        return len(self.data)  # O(1)
